src/api/crds.rs

```diff
@@ -0,0 +1,183 @@
+
+/// A data equivalent of the Resource trait for for Custom Resources
+///
+/// This is the smallest amount of info we need to run the API against a CR
+/// The version, and group must be set by the user.
+pub struct CustomResource {
+    kind: String,
+    group: String,
+    version: String,
+    api_version: String,
+    namespace: Option<String>,
+}
+
+impl CustomResource {
+    /// Construct a CrBuilder
+    pub fn kind(kind: &str) -> CrBuilder {
+        CrBuilder::kind(kind)
+    }
+}
+
+/// A builder for CustomResource
+#[derive(Default)]
+pub struct CrBuilder {
+    pub(crate) kind: String,
+    pub(crate) version: Option<String>,
+    pub(crate) group: Option<String>,
+    pub(crate) namespace: Option<String>,
+}
+impl CrBuilder {
+    /// Create a CrBuilder
+    ///
+    /// ```
+    /// use kube::api::{CustomResource, Resource};
+    /// struct FooSpec {};
+    /// struct FooStatus {};
+    /// struct Foo {
+    ///     spec: FooSpec,
+    ///     status: FooStatus
+    /// };
+    /// let foos : Resource = CustomResource::kind("Foo") // <.spec.kind>
+    ///    .group("clux.dev") // <.spec.group>
+    ///    .version("v1")
+    ///    .into_resource();
+    /// ```
+    fn kind(kind: &str) -> Self {
+        assert!(to_plural(kind) != kind); // no plural in kind
+        assert!(is_pascal_case(&kind)); // PascalCase kind
+        Self {
+            kind: kind.into(),
+            ..Default::default()
+        }
+    }
+
+    /// Set the api group of a custom resource
+    pub fn group(mut self, group: &str) -> Self {
+        self.group = Some(group.to_string());
+        self
+    }
+
+    /// Set the api version of a custom resource
+    pub fn version(mut self, version: &str) -> Self {
+        self.version = Some(version.to_string());
+        self
+    }
+
+    /// Set the namespace of a custom resource
+    pub fn within(mut self, ns: &str) -> Self {
+        self.namespace = Some(ns.into());
+        self
+    }
+
+    // Consume the CrBuilder and build a CustomResource
+    pub fn build(self) -> CustomResource {
+        let version = self.version.expect("Crd must have a version");
+        let group = self.group.expect("Crd must have a group");
+        CustomResource {
+            api_version: format!("{}/{}", group, version),
+            kind: self.kind,
+            version,
+            group,
+            namespace: self.namespace,
+        }
+    }
+
+    // Consume the CrBuilder and convert to an Api object
+    pub fn into_api<K>(self, client: APIClient) -> Api<K> {
+        let crd = self.build();
+        Api {
+            client,
+            api: crd.into(),
+            phantom: PhantomData,
+        }
+    }
+
+    // Consume the CrBuilder and convert to a Resource object
+    pub fn into_resource(self) -> Resource {
+        let crd = self.build();
+        crd.into()
+    }
+}
+
+/// Make Resource useable on CRDs without k8s_openapi
+impl From<CustomResource> for Resource {
```

Comment:
They seem to match 1:1, so.. why not just have `CrBuilder` build a `Resource` directly? (Assuming that this is still worthwhile to keep.)

Reply:
I could've called it a `ResourceBuider` perhaps - might be more appropriate - but it's only really useful for custom resources, that was why. It's almost the same, it infers the `api_version` from `version` and `kind`.